### ISSUE TITLE
Setting slow connectivity on network makes Geckodriver fail

### ISSUE THREAD

**Issue by soulgalore** [mozilla/geckodriver#1903]
## System

* Version: 0.29.1
* Platform: Mac Os Big Sur 11.4
* Firefox: 89.0.2 
* Selenium:  4.0.0-beta.4

A Browsertime user [reported](https://github.com/sitespeedio/sitespeed.io/issues/3414) that when they set local connectivity to 4g (to emulate a slow connection) Firefox stops to work. 

To throttle the connection we use https://github.com/sitespeedio/throttle that's a wrapper on Mac for _pfctl_ . If you don't wanna run throttle you can run the commands directly:

```
sudo dnctl -q flush
sudo dnctl -q pipe flush
sudo pfctl -f /etc/pf.conf
sudo pfctl -E
sudo pfctl -d
sudo dnctl -q flush
sudo dnctl -q pipe flush
sudo dnctl pipe 1 config delay 0ms noerror
sudo dnctl pipe 2 config delay 0ms noerror
echo "dummynet in all pipe 1
dummynet out all pipe 2" | sudo pfctl -f -
sudo dnctl pipe 1 config bw 9000Kbit/s delay 42.5ms
sudo dnctl pipe 2 config bw 9000Kbit/s delay 42.5ms
sudo pfctl -E
```

or use the Network Link Conditioner if you have that installed.

To reproduce I run:
`bin/browsertime.js -b firefox https://gmail.com/ -n 1 --connectivity.engine throttle -c 4g`

What happens is that Firefox never navigates to the page and we reach a 60s timeout. Turning on the log it looks like this:

```
1624648826490	Marionette	INFO	Listening on port 57591
1624648826490	Marionette	DEBUG	Marionette is listening
1624648826834	Marionette	DEBUG	Accepted connection 0 from 127.0.0.1:57609
1624648826927	Marionette	DEBUG	Closed connection 0
1624648827200	Marionette	DEBUG	Accepted connection 1 from 127.0.0.1:57612
1624648827300	Marionette	DEBUG	Closed connection 1
1624648827575	Marionette	DEBUG	Accepted connection 2 from 127.0.0.1:57614
1624648827677	Marionette	DEBUG	Closed connection 2
1624648827956	Marionette	DEBUG	Accepted connection 3 from 127.0.0.1:57615
1624648828057	Marionette	DEBUG	Closed connection 3
1624648828333	Marionette	DEBUG	Accepted connection 4 from 127.0.0.1:57616
1624648828436	Marionette	DEBUG	Closed connection 4
1624648828707	Marionette	DEBUG	Accepted connection 5 from 127.0.0.1:57617
1624648828808	Marionette	DEBUG	Closed connection 5
1624648829082	Marionette	DEBUG	Accepted connection 6 from 127.0.0.1:57618
1624648829184	Marionette	DEBUG	Closed connection 6
```

The log looks like the same problem as I had running Firefox at work on Android as in https://github.com/mozilla/geckodriver/issues/1817

Is there a way I can provide more/better logging?

**Comment by whimboo** [comment]
That's interesting. I wonder why the connection gets closed 100ms after it was accepted...

The [connection gets established](https://searchfox.org/mozilla-central/rev/c0fc8c4852e927b0ae75d893d35772b8c60ee06b/remote/marionette/server.js#158-161) and directly afterward we [run the handshake](https://searchfox.org/mozilla-central/rev/c0fc8c4852e927b0ae75d893d35772b8c60ee06b/remote/marionette/server.js#345-350). Transmitting the handshake should be fine because [otherwise geckodriver would raise an error](https://searchfox.org/mozilla-central/rev/c0fc8c4852e927b0ae75d893d35772b8c60ee06b/testing/geckodriver/src/marionette.rs#1114-1156).

Strangely your log output doesn't show anything from geckodriver. Could you please check why it's not there? It would be good to know if with debug logging the output for `Connection to Marionette established` would appear or not.

**Comment by soulgalore** [comment]
Hmm I'm not sure, if I enable verbose logging from Geckodriver it looks like this:
```
...
1627494618709	geckodriver::capabilities	DEBUG	Trying to read firefox version from ini files
1627494618710	geckodriver::capabilities	DEBUG	Found version 90.0.2
1627494618711	geckodriver::capabilities	DEBUG	Creating profile directory tree /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/chrome/
1627494618711	geckodriver::capabilities	DEBUG	Creating profile directory tree /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/extensions/
1627494618711	geckodriver::capabilities	DEBUG	Extracting profile to /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/extensions/browsertime-extension@sitespeed.io.xpi
1627494618712	geckodriver::capabilities	DEBUG	Extracting profile to /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/chrome/userChrome.css
1627494618712	geckodriver::capabilities	DEBUG	Extracting profile to /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/extensions/harexporttrigger@getfirebug.com.xpi
1627494618712	geckodriver::capabilities	DEBUG	Extracting profile to /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/cert9.db
1627494618713	geckodriver::capabilities	DEBUG	Extracting profile to /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/key4.db
1627494618723	mozrunner::runner	INFO	Running command: "/Applications/Firefox.app/Contents/MacOS/firefox-bin" "--marionette" "-devtools" "-no-remote" "-foreground" "-profile" "/var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57"
1627494618724	geckodriver::marionette	DEBUG	Waiting 60s to connect to browser on 127.0.0.1:54297
1627494619090	Marionette	INFO	Marionette enabled
1627494619163	Marionette	TRACE	Received observer notification toplevel-window-ready
2021-07-28 19:50:19.655 firefox-bin[11496:161387] Warning: Expected min height of view: (<NSButton: 0x128193c00>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2021-07-28 19:50:19.656 firefox-bin[11496:161387] Warning: Expected min height of view: (<NSButton: 0x13363d400>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2021-07-28 19:50:19.656 firefox-bin[11496:161387] Warning: Expected min height of view: (<NSButton: 0x13363a800>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2021-07-28 19:50:19.657 firefox-bin[11496:161387] Warning: Expected min height of view: (<NSButton: 0x13363dc00>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2021-07-28 19:50:19.657 firefox-bin[11496:161387] Warning: Expected min height of view: (<NSButton: 0x133637000>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2021-07-28 19:50:19.658 firefox-bin[11496:161387] Warning: Expected min height of view: (<NSPopoverTouchBarItemButton: 0x133636c00>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2021-07-28 19:50:19.658 firefox-bin[11496:161387] Warning: Expected min height of view: (<NSPopoverTouchBarItemButton: 0x1336cf400>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
console.warn: SearchSettings: "get: No settings file exists, new profile?" (new NotFoundError("Could not open the file at /var/folders/27/xpnvcsbs0nlfbb4qq397z3rh0000gn/T/rust_mozprofilelvak57/search.json.mozlz4", (void 0)))
console.error: BackgroundUpdate: 
  _reasonsToNotScheduleUpdates: Failed to check for Maintenance Service Registry Key: [Exception... "Component returned failure code: 0x80004001 (NS_ERROR_NOT_IMPLEMENTED) [nsIUpdateProcessor.getServiceRegKeyExists]"  nsresult: "0x80004001 (NS_ERROR_NOT_IMPLEMENTED)"  location: "JS frame :: resource://gre/modules/BackgroundUpdate.jsm :: _reasonsToNotScheduleUpdates :: line 243"  data: no]
1627494620932	Marionette	TRACE	Received observer notification marionette-startup-requested
1627494620932	Marionette	TRACE	Waiting until startup recorder finished recording startup scripts...
1627494620939	Marionette	TRACE	All scripts recorded.
1627494620940	Marionette	DEBUG	Setting recommended pref apz.content_response_timeout to 60000
1627494620940	Marionette	DEBUG	Setting recommended pref browser.contentblocking.introCount to 99
1627494620940	Marionette	DEBUG	Setting recommended pref browser.download.panel.shown to true
1627494620940	Marionette	DEBUG	Setting recommended pref browser.newtabpage.enabled to false
1627494620941	Marionette	DEBUG	Setting recommended pref browser.search.update to false
1627494620941	Marionette	DEBUG	Setting recommended pref browser.tabs.disableBackgroundZombification to false
1627494620942	Marionette	DEBUG	Setting recommended pref browser.tabs.remote.separatePrivilegedContentProcess to false
1627494620942	Marionette	DEBUG	Setting recommended pref browser.tabs.unloadOnLowMemory to false
1627494620942	Marionette	DEBUG	Setting recommended pref browser.tabs.warnOnCloseOtherTabs to false
1627494620942	Marionette	DEBUG	Setting recommended pref browser.tabs.warnOnOpen to false
1627494620942	Marionette	DEBUG	Setting recommended pref browser.toolbars.bookmarks.visibility to never
1627494620942	Marionette	DEBUG	Setting recommended pref browser.usedOnWindows10.introURL to 
1627494620942	Marionette	DEBUG	Setting recommended pref browser.urlbar.suggest.searches to false
1627494620943	Marionette	DEBUG	Setting recommended pref datareporting.policy.dataSubmissionPolicyAccepted to false
1627494620943	Marionette	DEBUG	Setting recommended pref dom.disable_beforeunload to true
1627494620943	Marionette	DEBUG	Setting recommended pref dom.file.createInChild to true
1627494620943	Marionette	DEBUG	Setting recommended pref extensions.getAddons.cache.enabled to false
1627494620943	Marionette	DEBUG	Setting recommended pref network.http.prompt-temp-redirect to false
1627494620943	Marionette	DEBUG	Setting recommended pref security.notification_enable_delay to 0
1627494620943	Marionette	DEBUG	Setting recommended pref signon.autofillForms to false
1627494620944	Marionette	DEBUG	Setting recommended pref signon.rememberSignons to false
1627494620945	Marionette	INFO	Listening on port 54297
1627494620945	Marionette	DEBUG	Marionette is listening
1627494626678	Marionette	DEBUG	Accepted connection 0 from 127.0.0.1:54298
1627494626769	Marionette	DEBUG	Closed connection 0
1627494627231	Marionette	DEBUG	Accepted connection 1 from 127.0.0.1:54308
1627494627333	Marionette	DEBUG	Closed connection 1
1627494628073	Marionette	DEBUG	Accepted connection 2 from 127.0.0.1:54309
1627494628111	Marionette	DEBUG	Closed connection 2
1627494628494	Marionette	DEBUG	Accepted connection 3 from 127.0.0.1:54313
1627494628609	Marionette	DEBUG	Closed connection 3
1627494629150	Marionette	DEBUG	Accepted connection 4 from 127.0.0.1:54318
...
```

**Comment by whimboo** [comment]
Thanks @soulgalore. I finally had a chance to take a look and inspected the geckodriver code. I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1723451 to get it fixed in the upcoming 0.30.0 release of geckodriver.

**Comment by whimboo** [comment]
I pushed a try build with a timeout of 1000ms. geckodriver builds should be available soon. @soulgalore maybe you could check if that fixes your problem? Thanks a lot

https://treeherder.mozilla.org/jobs?repo=try&revision=05d14687b19259c31ed756b6c5a0a7dca34d00ff&searchStr=gd-s

**Comment by soulgalore** [comment]
@whimboo yes I can try, how do I get the artefact from treeherder (where should I click), I haven't used it before.

**Comment by whimboo** [comment]
You simply open the above link and click on the green job name appropriate to your platform. In the lower pane you can find the download link of the signed geckodriver artifact.

**Comment by soulgalore** [comment]
@whimboo thanks! I verified and the updated version fixes the problem! Thank you :)

**Comment by whimboo** [comment]
Thanks @soulgalore! Do you still have logs from such a session? I would be interested in by using which network read timeout the handshake actually passes.

We actually decided to not increase the read timeout for each try to do a Marionette handshake, but bump it up from 100ms to 1000ms, which should be basically enough for each and every connection. Getting just a single byte within a second, should never actually happen. As such would you mind testing the following try build of geckodriver?

https://treeherder.mozilla.org/jobs?repo=try&revision=4ef709325db8beaa0a3c1cd3bf83c4f86fff0fad&searchStr=gd

**Comment by soulgalore** [comment]
Hi @whimboo I tried that build, that one works too, thanks! 

>  Do you still have logs from such a session? I would be interested in by using which network read timeout the handshake actually passes.

I didn't fully get that, one kind of setup do you need log from? I tried now with 3g settings (round trip time of 150 ms) and that worked. Trying with a really slow (round trip time 400 ms) made the latest build fail.

**Comment by soulgalore** [comment]
That trace looks like this:
```
1628587545359	Marionette	INFO	Listening on port 49295
1628587545359	Marionette	DEBUG	Marionette is listening
1628587548032	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
1628587549107	Marionette	DEBUG	Accepted connection 0 from 127.0.0.1:49309
1628587549473	Marionette	DEBUG	Closed connection 0
1628587551644	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
1628587552437	Marionette	DEBUG	Accepted connection 1 from 127.0.0.1:49318
1628587552438	Marionette	DEBUG	Closed connection 1
1628587554496	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
1628587555381	Marionette	DEBUG	Accepted connection 2 from 127.0.0.1:49329
1628587556864	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
1628587557521	Marionette	DEBUG	Accepted connection 1 from 127.0.0.1:49288
1628587558071	Marionette	DEBUG	Closed connection 1
1628587558373	Marionette	DEBUG	Closed connection 2
1628587559826	webdriver::server	DEBUG	-> GET /status 
1628587559826	webdriver::server	DEBUG	<- 200 OK {"value":{"message":"","ready":true}}
1628587560153	Marionette	DEBUG	Accepted connection 2 from 127.0.0.1:49336
1628587560283	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
1628587561483	Marionette	DEBUG	Closed connection 2
1628587562987	Marionette	DEBUG	Accepted connection 3 from 127.0.0.1:49345
1628587563128	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
1628587565730	Marionette	DEBUG	Closed connection 3
1628587569824	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
1628587573412	Marionette	DEBUG	Accepted connection 4 from 127.0.0.1:49351
1628587573424	Marionette	DEBUG	Closed connection 4
1628587585951	geckodriver::marionette	TRACE	Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 35). Retrying in 100ms
```

For our use case I've changed so we do not slow down the traffic on localhost. For the use case I had at work running on Android I have no way to test the new Geckodriver but my guess is that it fixed the problem.

**Comment by whimboo** [comment]
As discussed on Matrix the issues from the last comment were caused by a bug in throttle, which also caused a slowdown for `localhost` on MacOS. As such we are going forward and land my patch that increases the timeout from 100ms to 1000ms.

**Comment by soulgalore** [comment]
Sounds great!

**Comment by whimboo** [comment]
This got already fixed in geckodriver 0.30.0.
